### PR TITLE
Added user authentication.

### PR DESCRIPTION
**Pull request by innesian**
Added user authentication and hooked up the login form.

Push requires a resync of your database table structure and `npm install` on both server and client.

- Changed models to use 'underscores' for created_at and updated_at. Need these like this for the lando sync script.
- (Lando Config) Modified the database bash script to seed test data for Projects, Skills, Volunteers, Volunteer Skills. Added associations. Left a few volunteers without an associated project. 
- Pluralized "skill" on FE and BE hook.
- Modified the Lando sync script to push a user into the admin table. Credentials are voma.code.for.chicago@gmail.com with the password "password".
- Added express-session plugin and configured it for development.
- Refactored the login function in the admins controller to work with the login form. Just authentication is set atm we can add user type when those are figured out and ready.
- Stubbed out loginState function, this will get hit on page load and force a logout for unauthenticated users and reset the max-age of the session cookie.
- Added login function to the AuthProvider. On success redirect to `/board`
- Added error messaging to the Login form.
- Locked routes in Express that are not public.
- Setup proxy middleware in the client so that cookies from port 5000's server would set on the port 3000 frontend.

**Testing with Lando config:**
- Pull down, run `npm install` then wipe your box and rebuild it.
- `lando destroy` then `lando start` then `bash .lando.database.sh`
- Turn on the server and client. The credentials for an admin user should have been added by the database bash script.  Credentials are voma.code.for.chicago@gmail.com with the password "password". 

**Testing with other Databse Setup:**
- Pull down, run `npm install` then start the client and server.
- Add and admin user to the table, this is the correct hash for the password "password" .. `$2b$10$tNLcr6nHUm30/JEquBjCbuii77MHx4qXqjUc8JRRx0MnSW6FXT.BK`

### CODE REVIEW

**Comment by briwagner** [comment]
Looks like a lot of separate pieces here. It may be helpful if the changes were more atomic, so they could be tracked to work tickets, etc., and reviewed and tested.

I'm unclear on the Lando integration here; I don't see that we're recommending it for the dev environments. I see you're using it, Adam. Are others? If this is part of the dev workflow you're using, that's great. I don't know that those pieces should be part of the project overall though.

From the shell script, I don't see the need for modifying the models at this point. For seeding data, it would be possible with the sequelize-cli, similar to the recent migrations files. That was the hope for adding it to the project.

**Comment by innesian** [comment]
Sorry, definitely not tying to step on your toes with the config. When the sequelize-cli stuff is ready I'll be the first to use that. Problem is, I'm caught up to the assignment board on the frontend and need a seeded database to continue working on it. The database seeding script is not meant as an alternative to sequelize-cli it's a band-aid in the meantime so that setup isn't a blocker on the FE.

> I'm unclear on the Lando integration here; I don't see that we're recommending it for the dev environments.

Lando is a wrapper around Docker which is easy to install, I added it to the Readme as an alternative backend setup so it's being recommended there. It dumbs down the 11 step setup some of the volunteers have been struggling with. Literally, they only need to type two commands into a terminal to spin up Docker with a seeded database. I think we should nix Option A in the setup instructions and just use the Lando config until sequelize-cli is ready and polished.

**Comment by briwagner** [comment]
Lando is a cool tool. I've used if before. My hesitation is whether it's right for this project, assuming we have devs who are unfamiliar. I realize it sits on top of Docker and that's my concern. If folks are unfamiliar with Docker, they're likely to struggle with Lando as well. Even installing these tools may be an issue.

I defer to Will on this decision. But my two cents is: I don't suggest modifying the BE dev setup, or moving to Lando.

It's not clear to me where those folks may be struggling. I haven't worked with anyone in that situation on this project, and I don't want to jump to any conclusions. From what I saw with onboarding, the notion was that new devs would get the creds and just start working with a hosted DB somewhere. That could be the best way to proceed, to avoid having front-end devs have to manage _any_ DBs.

As far as this PR, I think it best if some of this can be broken up into more than one, maybe to match the issues. Adding express session is one. Changing "skill" to "skills" is another -- thanks for that. It would be easier to review and test and deploy, etc.

Sequelize-cli is ready to go. That was the PR you merged a couple days ago. Running the migration will give you the skills and projects. If there's a need for other seed tasks for dev, you could create those. Not sure what is needed at this point.

File: client/src/components/LoginForm.jsx
Comment (raindogg) [inline client/src/components/LoginForm.jsx:90]: Could you add this inside the `form` tag and have the event be triggered by `onSubmit` instead of `onClick`? That will allow users to submit the form by hitting enter and a bunch of other accessibility stuff.

**Review by raindogg — CHANGES_REQUESTED** [review]
A couple comments, but looks great! Thanks for your hard work on this. Just make sure to pull the latest branch of `main` before merging.

File: client/src/setupProxy.js
Comment (innesian) [reply client/src/setupProxy.js:1]: Good question, I spent a bit debugging this. Proxy (just set in the package.json) file is working as a proxy. Setting cookies from the proxied host doesn't seem to work for me. `express-session` sets a cookie `connect.sid` to identify a session/user and it's not setting for me without the middleware.

Maybe someone else can test on their machine to confirm. 

**To test:**
Check cookies with Chrome inspector, Application tab > Storage > Cookies.

- Just uninstall `http-proxy-middleware` and delete the `setupProxy.js` file, clear your node cache for good measure and in incognito mode verify you have no `connect.sid` cookie set for localhost:3000. 
- You'll hit the board page probably but check your cookies again and see if the express-session cookie is there.

File: client/src/components/LoginForm.jsx
Comment (innesian) [reply client/src/components/LoginForm.jsx]: Done.

**Review by raindogg — APPROVED** [review]

**Pull request merged**
